File: app/src/main/java/com/tamagotchi/android/tamagotchiAndroid/services/FloatingViewService.java
package com.tamagotchi.android.tamagotchiAndroid.services;

import android.app.Service;
import android.content.Intent;
import android.graphics.PixelFormat;
import android.os.IBinder;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.Toast;
import com.tamagotchi.android.tamagotchiAndroid.R;

public class FloatingViewService extends Service implements View.OnClickListener {
  private WindowManager mWindowManager;
  private View mFloatingView;
  private View collapsedView;
  private View expandedView;

  Animation movingAnim;

  public FloatingViewService() {
  }

  @Override
  public IBinder onBind(Intent intent) {
    return null;
  }

  @Override
  public void onCreate() {
    super.onCreate();

    // getting the widget layout from xml using layout inflater
    mFloatingView = LayoutInflater.from(this).inflate(R.layout.layout_floating_widget, null);

    // setting the layout parameters
    final WindowManager.LayoutParams params = new WindowManager.LayoutParams(
      WindowManager.LayoutParams.WRAP_CONTENT,
      WindowManager.LayoutParams.WRAP_CONTENT,
      WindowManager.LayoutParams.TYPE_PHONE,
      WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
      PixelFormat.TRANSLUCENT
    );

    // getting windows services and adding the floating view to it
    mWindowManager = (WindowManager) getSystemService(WINDOW_SERVICE);
    mWindowManager.addView(mFloatingView, params);

    //getting the collapsed and expanded view from the floating view
    collapsedView = mFloatingView.findViewById(R.id.layoutCollapsed);
    expandedView = mFloatingView.findViewById(R.id.layoutExpanded);

    //adding click listener to close button and expanded view
    mFloatingView.findViewById(R.id.buttonClose).setOnClickListener(this);
    expandedView.setOnClickListener(this);

    movingAnim = AnimationUtils.loadAnimation(this, R.anim.animation);
    movingAnim.setAnimationListener(new Animation.AnimationListener() {
      @Override
      public void onAnimationStart(Animation animation) {

      }

      @Override
      public void onAnimationEnd(Animation animation) {
        Toast.makeText(getApplicationContext(), "애니메이션 종료됨.", Toast.LENGTH_LONG).show();
      }

      @Override
      public void onAnimationRepeat(Animation animation) {

      }
    });

    // todo: this should be other types of event listener
//        collapsedView.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View view) {
//                collapsedView.startAnimation(movingAnim);
//            }
//        });

    // adding an touchlistener to make drag movement of the floating widget
    mFloatingView.findViewById(R.id.relativeLayoutParent).setOnTouchListener(new View.OnTouchListener() {
      private int initialX;
      private int initialY;
      private float initialTouchX;
      private float initialTouchY;

      @Override
      public boolean onTouch(View v, MotionEvent event) {
        switch (event.getAction()) {
          case MotionEvent.ACTION_DOWN:
            initialX = params.x;
            initialY = params.y;
            initialTouchX = event.getRawX();
            initialTouchY = event.getRawY();
            return true;

          case MotionEvent.ACTION_UP:
            //when the drag is ended switching the state of the widget
            collapsedView.setVisibility(View.GONE);
            expandedView.setVisibility(View.VISIBLE);
            return true;

          case MotionEvent.ACTION_MOVE:
            //this code is helping the widget to move around the screen with fingers
            params.x = initialX + (int) (event.getRawX() - initialTouchX);
            params.y = initialY + (int) (event.getRawY() - initialTouchY);
            mWindowManager.updateViewLayout(mFloatingView, params);
            return true;
        }
        return false;
      }
    });
  }

  @Override
  public void onDestroy() {
    super.onDestroy();
    if (mFloatingView != null) mWindowManager.removeView(mFloatingView);
  }

  @Override
  public void onClick(View v) {
    switch (v.getId()) {
      case R.id.layoutExpanded:
        //switching views
        collapsedView.setVisibility(View.VISIBLE);
        expandedView.setVisibility(View.GONE);
        break;

      case R.id.buttonClose:
        //closing the widget
        stopSelf();
        break;
    }
  }
}

File: app/src/main/java/com/tamagotchi/android/tamagotchiAndroid/services/FloatingViewService.kt
package com.tamagotchi.android.tamagotchiAndroid.services

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.graphics.PixelFormat
import android.os.Build
import android.util.Log
import android.view.*
import android.view.animation.Animation
import com.tamagotchi.android.tamagotchiAndroid.R
import android.view.GestureDetector.SimpleOnGestureListener
import com.tamagotchi.android.tamagotchiAndroid.views.ItemWidget
import com.tamagotchi.android.tamagotchiAndroid.views.MainWidget
import com.tamagotchi.android.tamagotchiAndroid.views.UpdownWidget

class FloatingViewService : Service() {
  val windowManager by lazy {
    getSystemService(WINDOW_SERVICE) as WindowManager
  }

  var floatingView: View? = null
  var buttonClose: View? = null

  override fun onBind(intent: Intent?): IBinder? = null

  override fun onCreate() {
    super.onCreate()
    init()
  }

  fun init() {
    MainWidget(this, windowManager)
    ItemWidget(this, windowManager)
    UpdownWidget(this, windowManager)
  }

  override fun onDestroy() {
    super.onDestroy()
    floatingView?.let {
      windowManager.removeView(it)
    }
  }
}

//setAnimationListener 코드
/*
  movingAnim = AnimationUtils.loadAnimation(this, R.anim.animation)
  movingAnim?.setAnimationListener(object : Animation.AnimationListener {
    override fun onAnimationStart(animation: Animation) {

    }

    override fun onAnimationEnd(animation: Animation) {
      Toast.makeText(applicationContext, "애니메이션 종료됨.", Toast.LENGTH_LONG).show()
    }

    override fun onAnimationRepeat(animation: Animation) {
    }
  })
*/
